File: 3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = (s) => {
  const previousChars = new Set();
  let left = 0;
  let right = 0;
  let count = 0;
  let maxCount = 0;

  while (right < s.length) {
    if (!previousChars.has(s[right])) {
      previousChars.add(s[right]);
      right++;
      count++;
      maxCount = Math.max(maxCount, count);
    } else {
      previousChars.delete(s[left]);
      left++;
      count--;
    }
  }

  return maxCount;
};

const s = "abcabcbb";
console.log('lengthOfLongestSubstring: ', lengthOfLongestSubstring(s))

File: 3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring.py
class Solution:
  def lengthOfLongestSubstring(self, s: str) -> int:
    previousChars = set()
    count = 0
    maxCount = 0
    left, right = 0, 0

    while right < len(s):
      if s[right] not in previousChars:
        previousChars.add(s[right])
        right += 1
        count += 1
        maxCount = max(maxCount, count)
      else:
        previousChars.remove(s[left])
        count -= 1
        left += 1

    return maxCount

s = "abcabcbb"
solution = Solution()
print('lengthOfLongestSubstring: ', solution.lengthOfLongestSubstring(s))
